libs/utils/src/num_wrapper.rs

```diff
@@ -0,0 +1,477 @@
+use parity_scale_codec::{
+	Compact, CompactAs, CompactRef, Decode, Encode, EncodeAsRef, EncodeLike, HasCompact, Input,
+	MaxEncodedLen, Ref, WrapperTypeDecode, WrapperTypeEncode,
+};
+use scale_info::TypeInfo;
+use serde::{Deserialize, Serialize};
+use sp_arithmetic::traits::{
+	Bounded, CheckedAdd, CheckedDiv, CheckedMul, CheckedNeg, CheckedRem, CheckedShl, CheckedShr,
+	CheckedSub, IntegerSquareRoot, One, Saturating, Zero,
+};
+use sp_std::{
+	cmp::Ordering,
+	fmt::{self, Debug},
+	marker::PhantomData,
+	ops::{
+		Add, AddAssign, Deref, Div, DivAssign, Mul, MulAssign, Rem, RemAssign, Shl, Shr, Sub,
+		SubAssign,
+	},
+};
+
+/// Type that allows to create different typed numbers with the same inner
+/// type:
+///
+/// ```
+/// # use crate::num_wrapper::NumWrapper;
+///
+/// struct Id1;
+/// struct Id2;
+///
+/// type FooU64 = NumWrapper<u64, Id1>;
+/// type BarU64 = NumWrapper<u64, Id2>;
+/// ```
```

Comment:
FooU64 and BarU64 are different types for the compiler, and you can implement different traits to them.